File: OpenCVによる画像処理入門/第3版/第2刷/sampleprogram/8-4-2.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;
string win_src = "src";
string win_dst = "dst";

int main()
{
  string file_src = "src_bin_noise1000.png"; // 入力画像のファイル名
  string file_dst = "dst.png"; // 出力画像のファイル名
  Mat img_src = imread(file_src, IMREAD_GRAYSCALE); // 入力画像（グレースケール）の読み込み

  Mat img_dst;
  if(!img_src.data) {
    cout << "error" << endl;
    return -1;
  }

  // ウインドウ生成
  namedWindow(win_src, WINDOW_AUTOSIZE);
  namedWindow(win_dst, WINDOW_AUTOSIZE);

  Mat img_tmp;
  Mat element8 = (Mat_<uchar>(3, 3) << 1, 1, 1, 1, 1, 1, 1, 1, 1); // 8近傍
  // オープニング
  morphologyEx(img_src, img_tmp, MORPH_OPEN, element8, Point(-1, -1), 1);
  // クロージング
  morphologyEx(img_tmp, img_dst, MORPH_CLOSE, element8, Point(-1, -1), 1);

  imshow(win_src, img_src); // 入力画像を表示
  imshow(win_dst, img_dst); // 出力画像を表示
  imwrite(file_dst, img_dst); // 処理結果の保存

  waitKey(0); // キー入力待ち
  return 0;
}

File: OpenCVによる画像処理入門/第3版/第2刷/sampleprogram/8-4-2.py
import cv2
import math
import numpy as np

file_src = 'src_bin_noise1000.png'
file_dst = 'dst.png'

img_src = cv2.imread(file_src, cv2.IMREAD_GRAYSCALE)

cv2.namedWindow('src')
cv2.namedWindow('dst')

element8 = np.array([[1, 1, 1], [1, 1, 1], [1, 1, 1]], np.uint8)  # 8近傍
# オープニング
img_tmp = cv2.morphologyEx(img_src, cv2.MORPH_OPEN, element8)
# クロージング
img_dst = cv2.morphologyEx(img_tmp, cv2.MORPH_CLOSE, element8)

cv2.imshow('src', img_src)  # 入力画像を表示
cv2.imshow('dst', img_dst)  # 出力画像を表示
cv2.imwrite(file_dst, img_dst)  # 処理結果の保存

cv2.waitKey(0)  # キー入力待ち
cv2.destroyAllWindows()
